src/unpack.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use libc::timeval;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::fs;
+use std::io;
+use std::path::Path;
+use tar::Archive;
+
+/// Unpack the contents of tarball to the destination path
+pub fn unpack(input: Vec<u8>, destination: &Path) -> Result<()> {
+    let mut archive = Archive::new(input.as_slice());
+
+    if destination.exists() {
+        return Err(anyhow!(
+            "unpack destination {:?} already exists",
+            destination
+        ));
+    }
+
+    fs::create_dir_all(destination)?;
+
+    let mut dirs: HashMap<CString, [timeval; 2]> = HashMap::default();
+    for file in archive.entries()? {
+        let mut file = file?;
+        file.unpack_in(destination)?;
+
+        // tar-rs crate only preserve timestamps of files,
+        // symlink file and directory are not covered.
+        if file.header().entry_type().is_symlink() || file.header().entry_type().is_dir() {
+            let mtime = file.header().mtime()? as i64;
+
+            let atime = timeval {
+                tv_sec: mtime,
+                tv_usec: 0,
+            };
+            let path = CString::new(format!(
+                "{}/{}",
+                destination.display(),
+                file.path()?.display()
+            ))?;
+
+            let times = [atime, atime];
+
+            if file.header().entry_type().is_dir() {
+                dirs.insert(path, times);
+            } else {
+                let ret = unsafe { libc::lutimes(path.as_ptr(), times.as_ptr()) };
+                if ret != 0 {
+                    return Err(anyhow!(
+                        "change symlink file: {:?} utime error: {:?}",
+                        path,
+                        io::Error::last_os_error()
+                    ));
+                }
+            }
+        }
+    }
+
+    // Directory timestamps need update after all files are extracted.
+    for (k, v) in dirs.iter() {
+        let ret = unsafe { libc::utimes(k.as_ptr(), v.as_ptr()) };
+        if ret != 0 {
+            return Err(anyhow!(
+                "change directory: {:?} utime error: {:?}",
+                k,
+                io::Error::last_os_error()
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use filetime;
+    use std::fs::File;
+    use std::io::prelude::*;
+    use tempfile;
+
+    #[test]
+    fn test_unpack() {
```

Comment:
Given the comments about the current limitations of `tar-rs`, shouldn't this also test sym-links? Also, what about other file types like char+block devices? I'm just wondering if we need tests for those here or can we rely on `tar-rs` for those tests?

Reply:
The reason we did not test sym-links is current `tar-rs` crate packing feature will not preserve the file time info, we may need a sepeate tool to crate the test file, I have add it as TODO for sym-links and other file types.